File: server_elixir/test/afterlight/activities/photo_booth_test.exs
defmodule Afterlight.Activities.PhotoBoothTest do
  use ExUnit.Case, async: true

  alias Afterlight.Activities.PhotoBooth

  describe "apply_input/3" do
    test "decline excludes that person; start waits on pending or proceeds without them" do
      state = PhotoBooth.init_sim_state(slots: [0, 1, 2])
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "invite", "slots" => [0, 1, 2], "commitId" => 1})
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "accept", "commitId" => 2})
      {state, _} = PhotoBooth.apply_input(state, 1, %{"kind" => "accept", "commitId" => 3})
      {state, event} = PhotoBooth.apply_input(state, 2, %{"kind" => "decline", "commitId" => 4})
      assert event["payload"]["excluded"] == true
      assert state["acceptedSlots"] == [0, 1]
      assert state["declinedSlots"] == [2]

      {_state, event} = PhotoBooth.apply_input(state, 0, %{"kind" => "start", "commitId" => 5})
      assert event["type"] == "countdown"
      assert event["payload"]["subjects"]["slots"] == [0, 1]
      assert event["payload"]["subjects"]["includeTheaterMedia"] == false
      assert event["payload"]["subjects"]["includeChat"] == false
      assert event["payload"]["subjects"]["includeBystanders"] == false

      waiting = PhotoBooth.init_sim_state(slots: [0, 1, 2])
      {waiting, _} = PhotoBooth.apply_input(waiting, 0, %{"kind" => "invite", "slots" => [0, 1, 2], "commitId" => 1})
      {waiting, _} = PhotoBooth.apply_input(waiting, 0, %{"kind" => "accept", "commitId" => 2})
      {waiting, _} = PhotoBooth.apply_input(waiting, 1, %{"kind" => "accept", "commitId" => 3})
      {_waiting, held} = PhotoBooth.apply_input(waiting, 0, %{"kind" => "start", "commitId" => 4})
      assert held["type"] == "capture_wait"
      assert held["payload"]["reason"] == "pending"
    end

    test "departure during countdown excludes that person" do
      state = PhotoBooth.init_sim_state(slots: [0, 1, 2])
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "invite", "slots" => [0, 1, 2], "commitId" => 1})
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "accept", "commitId" => 2})
      {state, _} = PhotoBooth.apply_input(state, 1, %{"kind" => "accept", "commitId" => 3})
      {state, _} = PhotoBooth.apply_input(state, 2, %{"kind" => "accept", "commitId" => 4})
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "start", "commitId" => 5})
      {state, _} = PhotoBooth.apply_input(state, 2, %{"kind" => "depart", "commitId" => 6})
      assert state["roster"]["2"] == "departed"
      refute 2 in state["acceptedSlots"]
      assert PhotoBooth.capture_subjects(state)["slots"] == [0, 1]
    end
  end

  describe "step_simulation/3" do
    test "four poses then a local-only strip with no image bytes" do
      state = PhotoBooth.init_sim_state(slots: [0, 1])
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "accept", "commitId" => 1})
      {state, _} = PhotoBooth.apply_input(state, 1, %{"kind" => "accept", "commitId" => 2})
      {state, _} = PhotoBooth.apply_input(state, 0, %{"kind" => "start", "commitId" => 3})

      {state, _} = PhotoBooth.step_simulation(state, %{}, 180)
      assert state["status"] == "posing"

      {state, event} =
        Enum.reduce(1..4, {state, nil}, fn _, {st, _} ->
          PhotoBooth.step_simulation(st, %{}, 80)
        end)

      assert state["status"] == "ready"
      assert state["stripReady"] == true
      assert state["imageBytes"] == nil
      assert state["uploadRequested"] == false
      assert event["type"] == "strip_ready"
      assert event["payload"]["upload"] == false
      assert event["payload"]["download"]["localOnly"] == true
    end
  end
end

File: tests/photo-booth.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  PHOTO_BOOTH_POSE_COUNT,
  PHOTO_BOOTH_POSES,
  PHOTO_BOOTH_COUNTDOWN_MS,
  PHOTO_BOOTH_POSE_MS,
  PHOTO_BOOTH_DT,
  validatePhotoBoothInput,
  initPhotoBoothState,
  applyPhotoBoothInput,
  stepPhotoBoothSimulation,
  photoBoothCaptureSubjects,
  photoBoothLocalDownload,
} from '../shared/photoBoothModel.js';
import { PhotoBoothModule, captureBoothStrip } from '../src/activities/photoBooth.js';
import { hasActivityModule, unregisterActivityModule } from '../src/activities/registry.js';

function stepsFor(ms) {
  return Math.ceil(ms / (PHOTO_BOOTH_DT * 1000));
}

test('photo booth: invite is 2-3 slots and unknowns are rejected', () => {
  assert.equal(validatePhotoBoothInput(null).valid, false);
  assert.equal(validatePhotoBoothInput({ kind: 'invite', slots: [0] }).valid, false);
  assert.equal(validatePhotoBoothInput({ kind: 'invite', slots: [0, 1, 2, 3] }).valid, false);
  assert.equal(validatePhotoBoothInput({ kind: 'invite', slots: [0, 1, 2] }).valid, true);
  assert.equal(validatePhotoBoothInput({ kind: 'upload' }).valid, false);
});

test('photo booth: a decline excludes that person; capture waits or proceeds without them', () => {
  let state = initPhotoBoothState({ activeSlots: [0, 1, 2] });
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'invite', slots: [0, 1, 2] }));
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 1, { kind: 'accept' }));
  let event;
  ({ simState: state, event } = applyPhotoBoothInput(state, 2, { kind: 'decline' }));
  assert.equal(event.payload.excluded, true);
  assert.deepEqual(state.acceptedSlots, [0, 1]);
  assert.deepEqual(state.declinedSlots, [2]);

  const waiting = applyPhotoBoothInput(state, 0, { kind: 'start' });
  // No pending left and two accepted — proceed
  assert.equal(waiting.event.type, 'countdown');
  assert.deepEqual(waiting.event.payload.subjects.slots, [0, 1]);
  assert.equal(waiting.event.payload.subjects.includeTheaterMedia, false);
  assert.equal(waiting.event.payload.subjects.includeChat, false);
  assert.equal(waiting.event.payload.subjects.includeBystanders, false);

  // Pending participant: wait
  state = initPhotoBoothState({ activeSlots: [0, 1, 2] });
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'invite', slots: [0, 1, 2] }));
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 1, { kind: 'accept' }));
  const held = applyPhotoBoothInput(state, 0, { kind: 'start' });
  assert.equal(held.event.type, 'capture_wait');
  assert.equal(held.event.payload.reason, 'pending');
});

test('photo booth: a departure during countdown excludes that person', () => {
  let state = initPhotoBoothState({ activeSlots: [0, 1, 2] });
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'invite', slots: [0, 1, 2] }));
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 1, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 2, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'start' }));
  const left = applyPhotoBoothInput(state, 2, { kind: 'depart' });
  assert.equal(left.simState.roster[2], 'departed');
  assert.ok(!left.simState.acceptedSlots.includes(2));
  assert.deepEqual(photoBoothCaptureSubjects(left.simState).slots, [0, 1]);
});

test('photo booth: four poses then a local-only strip with no upload bytes', () => {
  let state = initPhotoBoothState({ activeSlots: [0, 1] });
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 1, { kind: 'accept' }));
  ({ simState: state } = applyPhotoBoothInput(state, 0, { kind: 'start' }));

  ({ simState: state } = stepPhotoBoothSimulation(state, {}, stepsFor(PHOTO_BOOTH_COUNTDOWN_MS)));
  assert.equal(state.status, 'posing');
  assert.equal(state.poseIndex, 0);
  assert.equal(PHOTO_BOOTH_POSES.length, PHOTO_BOOTH_POSE_COUNT);

  for (let i = 0; i < PHOTO_BOOTH_POSE_COUNT; i++) {
    ({ simState: state } = stepPhotoBoothSimulation(state, {}, stepsFor(PHOTO_BOOTH_POSE_MS)));
  }
  assert.equal(state.status, 'ready');
  assert.equal(state.stripReady, true);
  assert.equal(state.imageBytes, null);
  assert.equal(state.uploadRequested, false);
  const download = photoBoothLocalDownload(state);
  assert.equal(download.localOnly, true);
  assert.equal(download.upload, false);
  const subjects = photoBoothCaptureSubjects(state);
  assert.equal(subjects.includeTheaterMedia, false);
  assert.equal(subjects.includeChat, false);
  assert.equal(subjects.includeBystanders, false);
});

test('photo booth: offscreen capture only draws consented slots and never uploads', () => {
  const frames = [];
  const result = captureBoothStrip({
    subjects: { slots: [0, 1], includeTheaterMedia: false, includeChat: false, includeBystanders: false },
    poses: PHOTO_BOOTH_POSES,
    drawFrame: (ctx, { slot, pose, poseIndex }) => {
      frames.push({ slot, pose, poseIndex });
      ctx.fillStyle = '#123';
      ctx.fillRect(0, 0, 8, 8);
    },
    createCanvas: (w, h) => {
      const store = { width: w, height: h, fills: [] };
      return {
        width: w,
        height: h,
        getContext: () => ({
          fillStyle: '',
          fillRect: () => { store.fills.push(1); },
          drawImage: () => {},
        }),
        toDataURL: () => 'data:image/png;base64,AA==',
      };
    },
  });
  assert.equal(result.upload, false);
  assert.equal(result.localOnly, true);
  assert.ok(result.dataUrl.startsWith('data:image/'));
  assert.ok(frames.every((f) => f.slot === 0 || f.slot === 1));
  assert.equal(frames.length, PHOTO_BOOTH_POSE_COUNT * 2);
  assert.ok(!result.theaterMedia);
  assert.ok(!result.chat);
});

test('photo booth: module registers photo-booth', () => {
  assert.equal(typeof PhotoBoothModule.initialize, 'function');
  assert.equal(hasActivityModule('photo-booth'), true);
  unregisterActivityModule('photo-booth');
});
